### ISSUE TITLE
Service delay / retry support

### ISSUE THREAD

**Issue by asbjornenge** [linuxkit/linuxkit#2280]
Would be nice to have some sort of `delay` or `retry` support for services in the yaml config.

Ref. conversation on slack:

```
asbjornenge [1:43 PM] 
Hmz, can I haz `delay` or `retry` for services? I have service X that builds some network bridges and interfaces, and service Y that needs those before starting...? (edited)

[1:44] 
Or do I need to handle @ service level? Specifically I'm running `serf` and need a specific network interface to exists before starting serf. I would like to avoid wrapping serf in some wait & try service... (edited)

[1:47] 
well, `sleep 10 && serf` is the quickfix I guess :stuck_out_tongue:

dave-tucker [1:57 PM] 
run the service that creates bridges `onboot`?

asbjornenge [2:04 PM] 
Well, it also runs a daemon and never exits... also it takes a little while cause it needs to register over some REST api's etc. would it still makes sense?

dave-tucker [2:06 PM] 
if it never exits, no it wouldn't make sense

asbjornenge [2:06 PM] 
yeah :confused:

[2:07] 
I guess the `sleep` trick should work...

dave-tucker [2:07 PM] 
not sure what the common opinion is, but you could share a file. e.g `/var/run/foo` between the services.... `echo done > /var/run/foo` in one and read until `/var/run/foo = done` in the other

asbjornenge [2:08 PM] 
yeah... and just wrap `serf` in a little bash script that check for `done`... I'll try that :slightly_smiling_face: Thanks! (edited)

dave-tucker [2:10 PM] 
you should definitely open an issue on the repo though as a `retry` and/or restart behaviour in general would be fairly useful things to specify in the YAML. IMHO (edited)

asbjornenge [2:22 PM] 
:+1: I'll do that :slightly_smiling_face:
```

**Comment by justincormack** [comment]
You really need to explain your use case in more detail. For creating network interfaces I would expect that to be able to happen purely in `onboot` containers (that is what I am planning for wireguard); but you seem to require a service to create them, which seems odd to me. Maybe you could explain what is happening behind the scenes a bit more.

**Comment by asbjornenge** [comment]
@justincormack yeah, I guess my case is a bit of an oddity here. I use `ZeroTier` which runs a daemon and where you also need to authorize access to a specific network before you can use it's interface.

The idea behind the issue was rather to support some `delay` of `retry` functionality (similar to dockers `restart=always` flag?) as a more general acknowledgement that some services might need that.

However, it worked fine for me using a `file signalling` approach mentioned in the chat.

Feel free to close the issue if you don't think it's a good idea. I'm not married to it 😛 👍  Atleast the `file signalling` approach is "documented" in this issue should someone else come looking 😉

**Comment by justincormack** [comment]
I dont really think a generic delay or retry would be that useful for you and I really dislike polling via restarting a container; you really need events. Currently the filesystem is our best event mechanism, later we do plan on having sockets (not in the filesystem) probably too. So going to close this for now.
